src/NHibernate/Type/ComponentType.cs

```diff
@@ -534,7 +534,7 @@ public override bool IsModified(object old, object current, bool[] checkable, IS
 				return current != null;
 			}
 			object[] currentValues = GetPropertyValues(current, session);
-			object[] oldValues = (Object[]) old;
+			object[] oldValues = old is object[] ? (Object[]) old : GetPropertyValues(old, session);
```

Comment:
This is a double cast implementation, which should be avoided. And better apply the fix from Hibernate as pointed by @hazzik. (hibernate/hibernate-orm@550c2e38ff41003c178a273b21f038d4295a0b80).

Granted, this will fix another issue by the way: bad handling of empty components. (They should be considered equal to `null`, which current implementation (with or without your fix) does not handle.) So their own test (hibernate/hibernate-orm@667136c473219ade173b7b3dfea50a1f9304c1a9) should be imported, and this PR/issue title and description should be adjusted.

Reply:
Thank you for the comments. Will do.